subprojects/core/src/integTest/groovy/org/gradle/api/provider/ProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.provider
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Unroll
+
+import static org.gradle.util.TextUtil.normaliseFileSeparators
+
+class ProviderIntegrationTest extends AbstractIntegrationSpec {
+
+    private static final String OUTPUT_FILE_CONTENT = 'Hello World!'
+    File defaultOutputFile
+    File customOutputFile
+
+    def setup() {
+        defaultOutputFile = file('build/output.txt')
+        customOutputFile = file('build/custom.txt')
+    }
+
+    @Unroll
+    def "can create and use provider by custom task written as #language class"() {
+        given:
+        file("buildSrc/src/main/$language/MyTask.${language}") << taskTypeImpl
+        buildFile << """
+            task myTask(type: MyTask)
+        """
+
+        when:
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+
+        when:
+        buildFile << """
+             myTask {
+                enabled = provider(true)
+                outputFiles = provider(files("${normaliseFileSeparators(customOutputFile.canonicalPath)}"))
+            }
+        """
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+        customOutputFile.isFile()
+        customOutputFile.text == OUTPUT_FILE_CONTENT
+
+        where:
+        language | taskTypeImpl
+        'groovy' | customGroovyBasedTaskType()
+        'java'   | customJavaBasedTaskType()
+    }
+
+    def "can assign default values for provider properties"() {
+        given:
+        buildFile << """
+            task myTask(type: MyTask)
+
+            class MyTask extends DefaultTask {
+                private Provider<Boolean> enabled = project.defaultProvider(Boolean)
+                private Provider<FileCollection> outputFiles = project.defaultProvider(FileCollection)
```

Comment:
You have exactly the same problem for a property of type `String`. Basically, we want a way to construct a `Provider<T>` that is live, that is, whose `get` method reflects changes to the underlying value. There are 2 ways to do this:

1. Return a live provider from the getter.
2. Add a separate API to query for an object's property as a `Provider<T>`.

For the first approach, one option is to store the current provider in the property and then create a live wrapper in the getter:

```
class MyExtension {
    Provider<String> someValue = defaultValue(String)
    FileCollection files = defaultValue(FileCollection)

    Provider<String> getSomeValue() { return provider { someValue } } // live wrapper
    void setSomeValue(String value) {  someValue = provider(value) }
    void setSomeValue(Provider<String> value) { someValue = value }

    FileCollection getFiles() { return provider { files } } // live wrapper
    void setFiles(FileCollection files) { this.files = files }
}

type(type: MyTask) {
    files = myExtension.files
    someValue = myExtension.someValue
}

myExtension.files = files("build/thing")
```

Another option is to store a mutable provider in the property and return this from the getter:

```
class MyExtension {
    // note: both fields are _final_ and always point to the same instance
    final PropertyState<String> someValue = property(String)
    final ConfigurableFileCollection files = property(ConfigurableFileCollection)

    Provider<String> getSomeValue() { return someValue }
    void setSomeValue(String value) { someValue.set(value) }
    void setSomeValue(Provider<String> value) { someValue.set(value) }

    FileCollection getFiles() { return files }
    void setFiles(FileCollection files) { this.files.set(files) }
}

type(type: MyTask) {
    files = myExtension.files
    someValue = myExtension.someValue
}

myExtension.files = files("build/thing")
```

This option has many advantages over the previous, because concerns such as immutability, state persistence and snapshotting, reclaiming memory, type conversion, validation, display name, provenance (where did this thing come from?) and so on can be taken care of automatically (by the mutable provider implementation).

For the second approach, again there are a few options:

One option is that the plugin author provides this:

```
class MyExtension {
    // using the 'mutable provider' but could also apply to the 'current provider' approach
    final PropertyState<String> someValue = property(String)
    final ConfigurableFileCollection files = property(ConfigurableFileCollection)

    String getSomeValue() { return someValue.get() }
    Provider<String> getSomeValueProvider() { return someValue }
    // plus setters    

    FileCollection getFiles() { return files }
    FileCollection getFilesProvider() { return files } // files is-a provider
    // plus setters 
}
```

Another option is that this is automatically provided:

```
class MyExtension {
    final PropertyState<String> someValue = property(String)
    final ConfigurableFileCollection files = property(ConfigurableFileCollection)

    String getSomeValue() { return someValue.get() }
    FileCollection getFiles() { return files }
}

type(type: MyTask) {
    // myExtension.properties (for example) is automatically provided for decorated object
    // provides access to a _live_ provider for each property
    someValue = myExtension.properties.someValue
    files = myExtension.properties.files
}

myExtension.files = files("build/out")
```